File: Monads/Examples.fs
module Monads.Examples

// F# часто використовує вбудовані монади: option, Result, async.
// Тут показано, як вони працюють, а також практичний приклад AsyncOption.
// Також додано власні оператори для зручності роботи з монадами.

let tryDivide x y : int option =
    if y = 0 then None else Some (x / y)

// Власні оператори для Option
let (>>=) m f = Option.bind f m
let (<!>) f m = Option.map f m

let optionExample () =
    let result: int option =
        Some 10
        >>= tryDivide 0
        >>= tryDivide 2
        >>= tryDivide 5
    match result with
    | Some value -> sprintf "Результат option: %d" value
    | None -> "Неможливо поділити"

let parsePositiveInt (text: string) : Result<int, string> =
    match System.Int32.TryParse(text) with
    | true, value when value > 0 -> Ok value
    | true, _ -> Error "Число має бути додатнім"
    | false, _ -> Error "Не вдалося розпарсити число"

// Власні оператори для Result
let (>>=!) m f = Result.bind f m
let (<!>) f m = Result.map f m

let resultExample () =
    let computation =
        parsePositiveInt "12"
        >>=! parsePositiveInt << string << (-) 2
        <!> (*) 3

    match computation with
    | Ok value -> sprintf "Результат Result: %d" value
    | Error msg -> sprintf "Помилка Result: %s" msg

let asyncWork x = async {
    do! Async.Sleep 100
    return x * x
}

let asyncExample () =
    async {
        let! a = asyncWork 3
        let! b = asyncWork 4
        return sprintf "Результат async: %d" (a + b)
    }

// Практичний AsyncOption: комбінація Async і option.
type AsyncOption<'T> = Async<option<'T>>

let asyncReturn x : AsyncOption<_> = async { return Some x }
let asyncZero<'T> : AsyncOption<'T> = async { return None }

let asyncBind (computation: AsyncOption<'T>) (binder: 'T -> AsyncOption<'U>) : AsyncOption<'U> = async {
    let! result = computation
    match result with
    | Some value -> return! binder value
    | None -> return None
}

let asyncMap (mapper: 'T -> 'U) (computation: AsyncOption<'T>) : AsyncOption<'U> = async {
    let! result = computation
    return Option.map mapper result
}

// Власні оператори для AsyncOption
let (>>=!!) m f = asyncBind m f
let (<!>!) f m = asyncMap f m

type AsyncOptionBuilder() =
    member _.Return(value) = asyncReturn value
    member _.Bind(computation, binder) = asyncBind computation binder
    member _.ReturnFrom(computation) = computation
    member _.Zero() = asyncZero

let asyncOption = AsyncOptionBuilder()

let tryDivideAsync x y : AsyncOption<int> = async {
    do! Async.Sleep 50
    if y = 0 then return None else return Some (x / y)
}

let asyncOptionExample () =
    async {
        let! result =
            asyncOption {
                let! a = tryDivideAsync 10 2
                let! b = tryDivideAsync a 5
                return a + b
            }

        return
            match result with
            | Some value -> sprintf "Результат AsyncOption: %d" value
            | None -> "Неможливо поділити асинхронно"
    }

let asyncOptionWithOperatorsExample () =
    async {
        let! result =
            asyncReturn 10
            >>=!! tryDivideAsync 2
            >>=!! tryDivideAsync 5
            <!>! id  // ідентична функція, щоб не змінювати значення

        return
            match result with
            | Some value -> sprintf "Результат AsyncOption з операторами: %d" value
            | None -> "Неможливо поділити асинхронно"
    }

let validateEven x : option<int> =
    Some x
    |> Option.bind (fun value -> if value % 2 = 0 then Some (value / 2) else None)

let tryDivideIfEven x : Result<int, string> =
    match validateEven x with
    | Some value -> Ok value
    | None -> Error "Число не парне"

let railwayExample () =
    let computation =
        parsePositiveInt "20"
        |> Result.bind tryDivideIfEven
        |> Result.map (fun x -> x + 5)

    match computation with
    | Ok value -> sprintf "Результат Railway: %d" value
    | Error msg -> sprintf "Помилка Railway: %s" msg

let parsePositiveAsync (text: string) : AsyncOption<int> =
    async {
        match parsePositiveInt text with
        | Ok value -> return Some value
        | Error _ -> return None
    }

let railwayAsyncExample () =
    async {
        let! result =
            asyncOption {
                let! value = parsePositiveAsync "20"
                let! half = tryDivideAsync value 2
                return half
            }
            |> asyncMap ((+) 1)

        return
            match result with
            | Some value -> sprintf "Результат Railway AsyncOption: %d" value
            | None -> "Помилка Railway AsyncOption"
    }

let runExamples () =
    printfn "%s" (optionExample())
    printfn "%s" (resultExample())
    asyncExample () |> Async.RunSynchronously |> printfn "%s"
    asyncOptionExample () |> Async.RunSynchronously |> printfn "%s"
    asyncOptionWithOperatorsExample () |> Async.RunSynchronously |> printfn "%s"
    railwayExample () |> printfn "%s"
    railwayAsyncExample () |> Async.RunSynchronously |> printfn "%s"

File: Monads/Examples.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Monads.Examples
{
    // У C# немає вбудованих монадів у термінах F#, але є практичні аналоги.
    // - Option: Nullable<T> або власний Maybe<T>
    // - Result: власний тип для обробки успіху/помилки
    // - Async: Task<T> та async/await
    // C# не підтримує створення власних операторів, тому використовуємо методи розширення.

    record Result<T, E>(T Value, E Error, bool IsOk)
    {
        public static Result<T, E> Ok(T value) => new(value, default!, true);
        public static Result<T, E> Error(E error) => new(default!, error, false);
    }

    static class ResultExtensions
    {
        public static Result<U, E> Bind<T, U, E>(this Result<T, E> result, Func<T, Result<U, E>> func) =>
            result.IsOk ? func(result.Value) : Result<U, E>.Error(result.Error);

        public static Result<U, E> Map<T, U, E>(this Result<T, E> result, Func<T, U> func) =>
            result.IsOk ? Result<U, E>.Ok(func(result.Value)) : Result<U, E>.Error(result.Error);
    }

    record Maybe<T>(T Value, bool HasValue)
    {
        public static Maybe<T> Some(T value) => new(value, true);
        public static Maybe<T> None => new(default!, false);
    }

    static class MaybeExtensions
    {
        public static Maybe<U> Bind<T, U>(this Maybe<T> maybe, Func<T, Maybe<U>> func) =>
            maybe.HasValue ? func(maybe.Value) : Maybe<U>.None;

        public static Maybe<U> Map<T, U>(this Maybe<T> maybe, Func<T, U> func) =>
            maybe.HasValue ? Maybe<U>.Some(func(maybe.Value)) : Maybe<U>.None;

        public static async Task<Maybe<U>> BindAsync<T, U>(this Task<Maybe<T>> maybeTask, Func<T, Task<Maybe<U>>> func)
        {
            var maybe = await maybeTask;
            return maybe.HasValue ? await func(maybe.Value) : Maybe<U>.None;
        }

        public static async Task<Maybe<U>> MapAsync<T, U>(this Task<Maybe<T>> maybeTask, Func<T, U> func)
        {
            var maybe = await maybeTask;
            return maybe.HasValue ? Maybe<U>.Some(func(maybe.Value)) : Maybe<U>.None;
        }
    }

    class Program
    {
        static int? TryDivide(int x, int y) => y == 0 ? null : x / y;

        static string OptionExample()
        {
            var result = TryDivide(10, 2)
                is int value ? TryDivide(value, 5) : null;

            return result.HasValue
                ? $"Результат option: {result.Value}"
                : "Неможливо поділити";
        }

        static Result<int, string> ParsePositiveInt(string text)
        {
            return int.TryParse(text, out var value)
                ? value > 0
                    ? Result<int, string>.Ok(value)
                    : Result<int, string>.Error("Число має бути додатнім")
                : Result<int, string>.Error("Не вдалося розпарсити число");
        }

        static string ResultExample()
        {
            var computation = ParsePositiveInt("12")
                .Bind(x => ParsePositiveInt((x - 2).ToString()))
                .Map(x => x * 3);

            return computation.IsOk
                ? $"Результат Result: {computation.Value}"
                : $"Помилка Result: {computation.Error}";
        }

        static async Task<string> AsyncExample()
        {
            async Task<int> AsyncWork(int x)
            {
                await Task.Delay(100);
                return x * x;
            }

            var a = await AsyncWork(3);
            var b = await AsyncWork(4);
            return $"Результат async: {a + b}";
        }

        static Task<Maybe<int>> TryDivideAsync(int x, int y) =>
            y == 0 ? Task.FromResult(Maybe<int>.None) : Task.FromResult(Maybe<int>.Some(x / y));

        static async Task<string> AsyncOptionExample()
        {
            var result = await TryDivideAsync(10, 2)
                .BindAsync(x => TryDivideAsync(x, 5));

            return result.HasValue
                ? $"Результат AsyncOption: {result.Value}"
                : "Неможливо поділити асинхронно";
        }

        static Result<int, string> ValidateEven(int x) =>
            x % 2 == 0
                ? Result<int, string>.Ok(x / 2)
                : Result<int, string>.Error("Число не парне");

        static string RailwayExample()
        {
            var computation = ParsePositiveInt("20")
                .Bind(ValidateEven)
                .Map(x => x + 5);

            return computation.IsOk
                ? $"Результат Railway: {computation.Value}"
                : $"Помилка Railway: {computation.Error}";
        }

        static Task<Maybe<int>> ValidateEvenAsync(int x) =>
            x % 2 == 0
                ? Task.FromResult(Maybe<int>.Some(x / 2))
                : Task.FromResult(Maybe<int>.None);

        static async Task<string> RailwayAsyncExample()
        {
            var result = await TryDivideAsync(20, 2)
                .BindAsync(x => ValidateEvenAsync(x))
                .MapAsync(x => x + 1);

            return result.HasValue
                ? $"Результат Railway AsyncOption: {result.Value}"
                : "Помилка Railway AsyncOption";
        }

        static async Task Main()
        {
            Console.WriteLine(OptionExample());
            Console.WriteLine(ResultExample());
            Console.WriteLine(await AsyncExample());
            Console.WriteLine(await AsyncOptionExample());
            Console.WriteLine(RailwayExample());
            Console.WriteLine(await RailwayAsyncExample());
        }
    }
}
